### ISSUE TITLE
Circular Dependency bundling 'folktale/result' into library with Rollup.

### ISSUE THREAD

**Issue by wking-io** [origamitower/folktale#180]
I am building a library that is using Result and when I am building a UMD bundle with Rollup I get Circular dependency errors

### Steps to reproduce

Minimal example here with steps:

https://github.com/wking-io/folktale-bug

#### Expected behaviour

I expected this to be bundled successfully without any errors.

#### Observed behaviour

There was an error because rollup found a Circular Depenedency

```
(!) Circular dependency: ../../node_modules/folktale/result/result.js -> ../../node_modules/folktale/conversions/result-to-validation.js -> ../../node_modules/folktale/validation/validation.js -> ../../node_modules/folktale/conversions/validation-to-result.js-> ../../node_modules/folktale/result/result.js
(!) Circular dependency: ../../node_modules/folktale/result/result.js -> ../../node_modules/folktale/conversions/result-to-validation.js -> ../../node_modules/folktale/validation/validation.js -> ../../node_modules/folktale/conversions/validation-to-result.js-> commonjs-proxy:/Users/wking/sites/saladbar/node_modules/folktale/result/result.js -> ../../node_modules/folktale/result/result.js
(!) Circular dependency: ../../node_modules/folktale/result/result.js -> ../../node_modules/folktale/conversions/result-to-validation.js -> ../../node_modules/folktale/validation/validation.js -> ../../node_modules/folktale/conversions/validation-to-maybe.js -> ../../node_modules/folktale/maybe/maybe.js -> ../../node_modules/folktale/conversions/maybe-to-result.js -> ../../node_modules/folktale/result/result.js
(!) Circular dependency: ../../node_modules/folktale/validation/validation.js -> ../../node_modules/folktale/conversions/validation-to-maybe.js -> ../../node_modules/folktale/maybe/maybe.js -> ../../node_modules/folktale/conversions/maybe-to-validation.js -> ../../node_modules/folktale/validation/validation.js
(!) Circular dependency: ../../node_modules/folktale/validation/validation.js -> ../../node_modules/folktale/conversions/validation-to-maybe.js -> ../../node_modules/folktale/maybe/maybe.js -> ../../node_modules/folktale/conversions/maybe-to-validation.js -> commonjs-proxy:/Users/wking/sites/saladbar/node_modules/folktale/validation/validation.js -> ../../node_modules/folktale/validation/validation.js
```

### Environment

(Describe the environment where the problem happens. This usually includes:

  - macOS
  - Node 8.9.4
  - Folktale 2.1.0
)

**Comment by Undistraction** [comment]
Note that this is potentially a bigger issue with Webpack 4's support of `sideEffects: false`. Packages with no side-effects (in the ES6 module sense), and (I think) no circular dependencies can declare themselves side-effect free by using this flag in their `package.json`, allowing Webpack to be much more aggressive in its tree-shaking (and potentially make huge savings on bundle size). I believe that Folktale's use of circular dependencies will preclude it, and possibly any lib that uses it from using this flag. I'm not 100% sure on this as info is a bit scarce at the moment, but I'll try and dig up some more information.

**Comment by robotlolita** [comment]
@Undistraction the idea of being able to add meta-data about runtime semantics that might not necessarily match the runtime semantics, then base optimisations off that gives me so many Common Lisp's optional typing optimisation vibes... It's not a good idea (what you get out of the compiler is unpredictable) :/

ES2015 modules' mutually recursive dependencies shouldn't affect tree-shaking, though. If it does, sounds like a bug in the library implementing the algorithm.

**Comment by Undistraction** [comment]
@robotlolita Yep. The whole thing sounds suspect to me and has the potential of nightmarishly hard to track-down bugs, but the savings in bundle size are hard to argue with. I haven't managed to track down a definitive answer but I'll post here if I do.

**Comment by newtriks** [comment]
Adding `external: ['folktale/validation'],` to my [Rollup config](https://rollupjs.org/guide/en#configuration-files) fixed the issue I had similar to this. The solution is specific to `validation` (maybe related to https://github.com/origamitower/folktale/issues/180) but easily tweaked to `result` I am sure...
